Migrate from twine upload to uv publish

John-P opened the pull request:
The `python.publish` task has been updated to use `uv publish` instead of `twine upload`.

Key changes:

- The `PublishTask` now constructs and executes a `uv publish` command.
- Authentication is handled via `UV_USERNAME` and `UV_PASSWORD` environment variables, which is `uv`'s standard practice.
- ~~A new `check_url` setting has been added to `PackageIndex` to support `uv`'s ability to check for existing packages against a different URL than the publishing repository.~~ This is now set in the call to uv using the index url (which was previously unused by the python publish task) when `skip_existing` is true.
- The `interactive` property has been ~~removed~~ deprecated.
  - `uv` has no interactive option, but will be interactive if not credentials are given.
- Tests and documentation have been updated to reflect the switch to `uv`.

John-P committed:
Replace Twine with uv for publishing packages

John-P committed:
Add missing init arg

John-P committed:
Markdown link formatting

John-P commented:
I see that uv is included as a dependency of Kraken. However, would a check and error message if uv is not in the `PATH` be useful? Also is a custom `uv_cmd` (as was used with twine previously) still needed? I removed this as I saw uv included in the dependencies and other code shelling out to uv without a variable cmd.

John-P commented:
> The unused interactive property has been removed. 

Perhaps keeping interactive but with a deprecation warning noting that it does nothing would be more helpful to not break things initially for users.

John-P commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -33,31 +33,28 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
         self.logger.info("$ %s", safe_command)
 
-        returncode = subprocess.call(command, cwd=self.project.directory)
+        returncode = subprocess.call(command, cwd=self.project.directory, env=env)
         return TaskStatus.from_exit_code(safe_command, returncode)
```
```suggestion
    try:
        returncode = subprocess.call(command, cwd=self.project.directory, env=env)
    except FileNotFoundError:
        return TaskStatus.failed(
            "The 'uv' command was not found. The python.publish task requires uv to be installed. "
            "See https://github.com/astral-sh/uv for installation instructions."
        )

        return TaskStatus.from_exit_code(safe_command, returncode)
```

NiklasRosenstein replied:
If we use the `uv` binary that comes with Kraken through it's dependencies, we wouldn't need that check. That UV version is pinned with `.kraken.lock`, which seems like a desirable property; although on the other hand it might mean the `uv` version run by users when invoked manually is likely a different one..

John-P commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -33,31 +33,28 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
 
     def execute(self) -> TaskStatus:
         credentials = self.index_credentials.get()
```
With
`interactive: Property[bool | None] = Property.default(None)`

```suggestion
    def execute(self) -> TaskStatus:
        # Check for the deprecated property
        if self.interactive.get() is not None:
            warnings.warn(
                "The 'interactive' property on the python.publish task is deprecated and has no effect. "
                "uv publish is non-interactive by default in this context.",
                DeprecationWarning,
            )
        credentials = self.index_credentials.get()
```

ulucs commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -66,12 +63,10 @@ def publish(
     group: str | None = "publish",
     project: Project | None = None,
     after: list[Task] | None = None,
-    twine_version: str = ">=6.0.0,<7.0.0",
```
can we print a warning here to warn the users that this argument has been removed and will be ineffective?

NiklasRosenstein replied:
Practically nobody is using that argument; according to internal search `kraken-hs` is the only reference for `twine_version` and it pins the minor version of `kraken-build` so this would be safe to remove.

ulucs commented on kraken-build/src/kraken/std/python/settings.py:
```diff
@@ -90,6 +94,8 @@ def add_package_index(
             If not specified, *alias* must be a known package index (`pypi` or `testpypi`).
         :param upload_url: If the upload url deviates from the registry URL. Otherwise, the upload URL will
             be the same as the *url*,
+        :param check_url: Optional index URL to check for existing files. Otherwise, the index URL will be
+            used.
```
Does this behavior come from `uv`? I would have expected to see `index_url` as the default value of `check_url`

John-P replied:
Yes this is from uv and the index_url is the default if check_url is not given to uv. We could explicitly assign check_url to index_url here if that would be more clear.

NiklasRosenstein replied:
Do we need the `check_url` configured separately? Is it because the URL cannot be the `index_url` for Artifactory? 

I would consider a `exist_ok` property on the publish task a cleaner way to express the intent.

John-P replied:
Sooo...I have looked into this a bit further and twine and uv actually differ in behavior here.

### They determine how to skip files differently

- **Twine** [relies on status codes (409)](https://github.com/pypa/twine/blob/974a8d8b1e63f65ab31936ae39a63adcdf3f47c6/twine/commands/upload.py#L32-L64) which is returned on existing same file name
- **uv** uses content-based hashes (assumes the registry supports SHA hashes in the [simple PyPI API](https://packaging.python.org/en/latest/specifications/simple-repository-api/#simple-repository-json-project-detail)). See the quote below from the uv docs.

### uv `--check_url` with PyPI vs Others

With uv, the `--check-url` option behaves differently for PyPI vs other registries. For PyPI this is not required, because the index URL for PyPI is known to uv (since it is the default registry).

From the [uv documentation](https://docs.astral.sh/uv/guides/package/#publishing-your-package) (emphasis added):

> Even though uv publish retries failed uploads, it can happen that publishing fails in the middle, with some files uploaded and some files still missing. **With PyPI, you can retry the exact same command, existing identical files will be ignored. With other registries, use --check-url <index url> with the index URL** (not the publishing URL) the packages belong to. **When using --index, the index URL is used as check URL**. uv will skip uploading files that are identical to files in the registry, and it will also handle raced parallel uploads. Note that **existing files need to match exactly** with those previously uploaded to the registry, this avoids accidentally publishing source distribution and wheels with different contents for the same version.

So in order for skipping existing files to work, either:
- The index URL must be set, OR
- The check URL must be set (usually the same as index-url). This is in case you want to use a different URL to check files to skip than is used for the regular index API.

### Additional Failures from Different Behaviour

Since uv determine files to skip based on the hash and not the filename, this means that an upload may fail when it would not for twine `--skip-existing`. For example, if the code was changed and an upload is attempted. With twine this would be ok, and files would simply be ignored (unless a new file is added) with no error. However, with uv this would be an error.

### Proposed Action

The index URL is a non-optional attribute `PackageIndex` but was/is unused. Assuming this is correctly set, we could just actually use this which will enable skipping uploading existing files. Then the check URL is not really needed, assuming that having a different check URL is quite uncommon. However, it is not difficult to support this.

Lastly, if you would like the behavior from twine of "I changed my code but didn't bump the version. Do not overwrite the same files and do not error" we would have to check the exit code on upload and ignore it when a flag is set.

NiklasRosenstein replied:
> Note that existing files need to match exactly with those previously uploaded to the registry, this avoids accidentally publishing source distribution and wheels with different contents for the same version.

For our use case (retrying a CI job that failed and partially uploaded some packages) this will likely be insufficient. The archive produced by `uv build` is not deterministic and the content hash of the generated packages will be different on a retry. Hence, it will try to upload _all_ files, and fail again but this time for the files already published.

> The index URL is a non-optional attribute PackageIndex but was/is unused.

When you say unused, you mean in the context of publishing packages? Because it is used, just elsewhere (e.g. to inject the indexex into the `pyproject.toml`). If the index URL is to be the same value as the check URL in practically all cases, I think we can forego a separate check URL.

Or alternatively, we can do both: Add a `check_url` that if set and supported will be used to check for existing packages, but the publish task also gets an `exist_ok` option, which will have the task make sure the same package getting published again is OK and not an error (in which case it will use the `check_url` if available or otherwise `index_url`, or by relying on the behavior of Uv to do that fallback).

But the semantics of an `exist_ok` or `skip_existing` property must be clear. The existing behavior with Twine was to accept conflict responses as you pointed out (409 Conflict), but Uv seems to rely on file hashes. Those are two very distinct semantics that will behave very differently. We might need to implement the check ourselves to recover Twine's `--skip-existing` behavior.

Note that we already implement a similar check for Cargo ourselves: https://github.com/kraken-build/kraken/blob/develop/kraken-build/src/kraken/std/cargo/tasks/cargo_publish_task.py#L153

John-P replied:
> When you say unused, you mean in the context of publishing packages? 

Yes, sorry if unclear.

Thanks for the clarification on indented behavior here. Looking into it

John-P replied:
I have now added a prepare check similar to cargo which will use the index URL and PyPI simple API to check for an uploaded file listing and will skip uploading if the distribution files already exist. This should emulate the behavior of the old twine `--skip-existing`.

I have tested this with the pypiserver and also manually with Artifactory (Which both return the HTML response format).

ulucs replied:
Noting here that our python builds are non-deterministic and this difference in behavior will show itself quite often. I am of the opinion that checking the hash and erroring is a desirable approach (even if it differs from what twine does), but I'm deferring to Nik on what Kraken should be doing.

NiklasRosenstein replied:
With the changes added by @John-P to ignore distribution files that are already in the index, I think this can be resolved.

NiklasRosenstein commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -33,31 +33,28 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
+        if credentials:
+            # See https://docs.astral.sh/uv/guides/package/#publishing-your-package
+            # NOTE: PyPI does not support publishing with username and password anymore,
+            #       should we log a warning if the username is not __token__?
```
I think that's reasonable! The check could be `"pypi.org/" in self.index_upload_url.get() and credentials[0] != "__token__"`. Would be good to include a bit of info as well like `logger.warning("since 2025-MM-DD, PyPi no longer allows publishing with username/password, see https://...")`

NiklasRosenstein commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -33,31 +33,28 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
         self.logger.info("$ %s", safe_command)
 
-        returncode = subprocess.call(command, cwd=self.project.directory)
+        returncode = subprocess.call(command, cwd=self.project.directory, env=env)
```
nit: would prefer this style, as it keeps the intent of "we need to pass the full environment not just overrides" more localised

```suggestion
        returncode = subprocess.call(command, cwd=self.project.directory, env={**os.environ, **env})
```

John-P replied:
env is initialized with os.environ so this would have no effect surely?

NiklasRosenstein reviewed (commented):
Left a few comments; looks good overall! 👍 Not approving just yet due to open discussions

John-P committed:
Add deprecation warning for interactive

John-P committed:
Add pypi token username warning

John-P committed:
Fix tests

John-P committed:
Fix --no-progress placement in command

John-P committed:
Keep deprecated interactive arg

John-P committed:
Update test

John-P committed:
Remove check_url

John-P committed:
Update test

John-P committed:
Move file args to end of uv upload

John-P committed:
Fix check-url conditional

John-P committed:
Update message

John-P commented:
Interesting to see that twine has removed support for skip-existing on non-pypi indexes: https://github.com/pypa/twine/pull/1253

John-P committed:
Init env as empty dict

John-P committed:
Check index for existing packages before publish

Refactor the `PublishTask` to proactively check the package index for existing distributions when `skip_existing` is true.

The redundant `exists_ok` property has been removed.

John-P committed:
Support both html content types

John-P committed:
Add test with mock data for html and json pypi simple api variants

Fix mypy error

John-P committed:
Refactor to use task logger instead of loguru

Logger fix

John-P committed:
Add missing import for pytest_httpx HTTPMock

John-P committed:
Remove comment

John-P committed:
Rename variable

John-P committed:
Typo

John-P committed:
Remove assert

John-P commented:
One thing I noticed while working on this is that there is a mix of loguru and logging used here which I was a bit confused about. I am using Task.logger here but also could not get this to output anywhere, so used loguru for testing/debugging. It looks like there is no handler set up? Just wanted to query if I was doing something incorrect there or if this is something that need fixing.

NiklasRosenstein commented:
Loguru was introduced new a while ago, but changing `Task.logger` to a Loguru logger would be a (potentially very silent) breaking change. I think using `Task.logger` is fine regardless where one has access to it, and otherwise `loguru`.

Did you try increasing the verbosity to get the logs to show?

The intercept handler is configured here:

https://github.com/kraken-build/kraken/blob/7bdab959c52e7fd789813094d2a742ce0423bcea/kraken-build/src/kraken/common/_option_sets.py#L59-L63

NiklasRosenstein commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -31,47 +36,166 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
         yield from (TaskRelationship(task, True, False) for task in self.dependencies)
         yield from super().get_relationships()
 
+    def _get_existing_files_from_index(self, project_name: str) -> set[str] | None:
```
I think this function would do well to live as a standalone function; I don't see a good reason for it to live as a member method. If nothing else, it could be a `@staticmethod` that can then also be tested independently of an instance of the Task (after adding parameters for the index URL and credentials).

John-P replied:
I have extracted it as a stand-alone function now. See https://github.com/kraken-build/kraken/pull/389/commits/cfa2acf71647aa2fbf112fc9b3e5c54910353ad9

NiklasRosenstein commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -31,47 +36,166 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
+            check=False,
+        )
+
+        if all(
```
Given the manual filtering of the distributions based on the existing files that already exist in the index, wouldn't we expect to never run into this condition?

John-P replied:
Ah yes, this *should* not longer be possible so I think it is now a bit redundant. See https://github.com/kraken-build/kraken/pull/389/commits/b47f973dbdcc3e6c433d993bec8afc18a1f49849

NiklasRosenstein reviewed (commented):
Left two more comments. :)

Very nice job btw., particularly on those tests! 🚀

John-P committed:
Refactor `_get_existing_files_from_index` as a standalone function

John-P committed:
Remove redundant uv stderr hash mismatch check

John-P committed:
Add tests for `_get_existing_files_from_index`

John-P committed:
Use f-strings for loguru messages

John-P commented:
Thanks for the review. I think I have addressed the comments on extracting the function and adding some tests for it also. Sorry for the slightly messy conversation log! I did a rebase to try and tidy up the git history after a minor mistake but did not realise it would make the conversation view a mess!

NiklasRosenstein commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -31,47 +37,157 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
+        existing_files = _get_existing_files_from_index(
+            project_name, self.index_index_url.get(), self.index_credentials.get()
+        )
+        logger.debug(f"Existing files: {existing_files}")
```
```suggestion
        self.logger.debug("Existing files in index: %s", existing_files)
```

For consistency

NiklasRosenstein commented on kraken-build/src/kraken/std/python/tasks/publish_task.py:
```diff
@@ -31,47 +37,157 @@ def get_relationships(self) -> Iterable[TaskRelationship]:
+            env["UV_PUBLISH_PASSWORD"] = credentials[1]
+
+        safe_command = command
         self.logger.info("$ %s", safe_command)
```
```suggestion
        safe_command = command
        self.logger.info("$ %s", command)
```

Since the command itself does not contain unsafe data

NiklasRosenstein reviewed (changes requested):
Almost ready  🚀

John-P committed:
Change from loguru to std logging in prepare

Co-authored-by: Niklas Rosenstein <rosensteinniklas@gmail.com>

John-P committed:
Adjust command log

Co-authored-by: Niklas Rosenstein <rosensteinniklas@gmail.com>

NiklasRosenstein committed:
flup to 5fa6a1c: Remove unused safe_command variable
